`directionalCombo`: handle `[key]=left|right|both` schema (Support `sidewalk=right`)

tordans opened the pull request:
This resolves https://github.com/openstreetmap/iD/issues/10839

Following what was done in https://github.com/openstreetmap/iD/pull/9587/files

---

This is ready for review, see https://github.com/openstreetmap/iD/pull/10935#issuecomment-2973569136

tordans committed:
`directionalCombo`: handle `[key]=left|right|both` schema

tordans commented:
@k-yle @tyrasd can you give me some pointers as to why the specs don't work and if this is generally a sane way to do this?

gy-mate commented:
@k-yle @tyrasd Could you please help @tordans with the failing test specs? As @waldyrious previously [pointed out](https://github.com/openstreetmap/id-tagging-schema/pull/1507#issuecomment-2779865864), it would be great to have this feature!

Many thanks in advance! :)

k-yle reviewed (commented):
@gy-mate we discussed this on slack, there's two lines that need to be added [to `onChange`](https://github.com/openstreetmap/iD/blob/e07dcddfee4c26e4f8e59b967cb0ac96bc619b5a/modules/ui/fields/directional_combo.js#L96) which will fix the tests and the ui bug:

```diff
+ if (otherValue === 'left') otherValue = otherKey.endsWith(':right') ? 'yes' : 'no';
+ if (otherValue === 'right') otherValue = otherKey.endsWith(':left') ? 'yes' : 'no';
```

(commenting here since I can't start a thread on unchanged lines)

tordans commented:
@k-yle thanks a lot, that got me on the right track, finally. 
Here is the current code with green specs:


https://github.com/user-attachments/assets/0ca9d731-2ebf-4fdc-a40f-1a49b9d6383d

IMO this is finished, now, and ready for a "ready to merge" review.

gy-mate commented:
Awesome, thank you so much! :)

tordans commented:
On a separate note, the specs are not as nice as I wanted them to be. Especially https://github.com/openstreetmap/iD/pull/10935/files#diff-f10edc7d646966248a759f9414d26ed695bf57d21db38f18b76f2146d8528d63R231-R243
I can only test left and right separately and check that one value was updated correctly but have to ignore the other values (see "ignore" comment). Ideally we had a check below that that would show that after triggering the update for each side, that both values are now updated. I tried duplicating https://github.com/openstreetmap/iD/pull/10935/files#diff-f10edc7d646966248a759f9414d26ed695bf57d21db38f18b76f2146d8528d63R225-R229 again for that but that does not work.

I think it's fine now, but curious about how to actually do it…

tordans committed:
Update changelogt: directional combo supports left|right|both)

tordans committed:
Merge branch 'develop' into directionalcombo-side-values

tordans committed:
Merge branch 'develop' into directionalcombo-side-values

tyrasd committed:
Merge branch 'develop' into directionalcombo-side-values

tyrasd committed:
fix logic of interpreting direction mapped as (common) tag values

To summarize:

* `sidewalk=left` should be interpreted as `sidewalk:left=yes` + `sidewalk:right=no` (and v.v. for `sidewalk=right`)
* `sidewalk=both` should be interpreted as `sidewalk:left=yes` + `sidewalk:right=yes`
* `sidewalk=other` should be interpreted as `sidewalk:left=other` + `sidewalk:right=other`
* `sidewalk:both=value` should always be interpreted as `sidewalk:left=value` + `sidewalk:right=value` (also when value=left/right)

This should always be the case, regardless if the preset uses `:both` for the common key or not.

Also needs to ensure correct handling of multiselections: The fallback mechanism between the ":both" and ":<direction>" tags of a multiselection requires to look at the tags of each selected entity individually: otherwise the undefined state of a specific direction of the field might be overwritten by the value of the ":both" tag on another selected entity. For example:

* entity 1 has `sidewalk:both=yes`
* entity 2 has `sidewalk:left=yes`

The multiselection now must represent that for the `right` direction, there are conflicting values: once 'no', once _unset/undefined_.

(see also 9423 / 4c222fde0a)

tyrasd reviewed (approved):
While testing, I found that the PR did not handle the main use case for which we need it (https://github.com/openstreetmap/id-tagging-schema/pull/1507) very gracefully. For example, when the preset uses `sidewalk:both` as the _common_ key, the transformation of `sidwalk=left` would not result in the fields to show it like as if it was `sidewalk:left/right=yes/no`. There were also some corner cases that were not handled correctly, e.g. a present tag of the `otherKey` should never be overwritten by the fallback mechanism (e.g. `sidewalk=left + sidewalk:right=separate` should not leave the `sidewalk:right` tag untouched when changing the _left_ value of the field).

Lastly, I also found some tricky edge cases for when a multi-selection is containing several OSM objects that use different tagging "variants" (e.g. one tagged as `sidewalk:left=yes` and once `sidewalk=both`). Strictly speaking, that was already a problem in the current code (e.g. reproducible when there is one `cycleway:left=lane` and one `cycleway=track`: it would show as if the multiselection of both had a _Bike Track_ on the right side of both and only conflicting values for the left side), but as this very much related to the code of this PR anyway, I also addressed it was well.

See e2cf8e979499f7c70b4bbac1a30e326baf4722fc (and 66d3e23f) for the updated logic. //edit PS: it also generalized the code a little further, as to not assume that `:left`/`:right` are the two allowed direction options (there could be new fields in the future using `:forward:`/`backward` for example) nor that the `:<direction>` part is always at the end of the tag key (there's already the `parking:<direction>:orientation` field for which this is not the case).

tyrasd committed:
also make tag-change event handler less hardcoded
